guides/source/asset_pipeline.md

```diff
@@ -8,479 +8,616 @@ This guide covers the asset pipeline.
+    [esbuild](https://esbuild.github.io/), or [Vite](https://vite.dev/) into
+    your Rails application, enabling you to handle complex asset requirements
+    alongside Propshaft.
+
+    Use `js-bundling-rails` with these tools to bundle JavaScript files, manage
+    dependencies, and transpile modern JavaScript for browser compatibility. For
+    more details, check the [`js-bundling-rails`
+    documentation](https://github.com/rails/jsbundling-rails).
+
+    Use `css-bundling-rails` to pre-process stylesheets using tools like
+    [Sass](https://sass-lang.com/) or [Tailwind CSS](https://tailwindcss.com/),
+    making it easier to work with advanced CSS features. For more details, check
+    the [`css-bundling-rails`
+    documentation](https://github.com/rails/cssbundling-rails).
+
+    To read more about how to use these gems you can see the [Advanced Asset
+    Management section](#advanced-asset-management). These tools complement
+    Propshaft by handling the complex processing, while Propshaft efficiently
+    organizes and serves the final assets.
+
+### Versioning with digest-based URLs
+
+In Rails, asset versioning helps manage cache behavior by adding unique
+identifiers to asset filenames. This ensures that whenever an asset is updated,
+the browser sees it as a new file and loads the latest version instead of using
+an older, cached one.
+
+Sprockets automatically handles asset fingerprinting by appending a hash to
+filenames whenever assets are updated. Similarly, Propshaft uses content-based
+fingerprinting and automatically updates asset paths in CSS to reference
+digested filenames. However, for JavaScript files, you’ll need to manually
+trigger the transformation or use a bundler to ensure the correct filenames are
+applied.
+
+You can read more about how to use asset versioning in the [Fingerprinting
+section](#fingerprinting).
+
+### Asset Precompilation
+
+In Propshaft, all assets from the paths configured in `config.assets.paths` are
+available for serving and will be copied into `public/assets` when precompiling.
+Unlike Sprockets, which only copied assets that were explicitly included in a
+bundle, Propshaft automatically processes all assets from the specified paths.
+This includes images, stylesheets, JavaScript files, and more.
+
+If you want to exclude certain directories from this process, you can use
+`config.assets.excluded_paths`. Read more about this in the [Fingerprinting
+section](#fingerprinting).
+
+Once precompiled, these assets can be referenced by their logical path using
+helpers like `asset_path`, `image_tag`, and `javascript_include_tag`. In production,
+when `assets:precompile` is run, these references are automatically turned into
+digest-aware paths. This is done using a `manifest.json` file that is
+automatically generated in` public/assets/.manifest.json`. This file maps logical
+asset paths to their precompiled file paths.
+
+Using Propshaft in your app
```

Comment:
This feels a little like burying the lede. For a Rails 8 guide, it feels like this should be:
- This is an asset pipeline
- Propshaft takes care of the asset pipeline
- These are Propshaft features and how to use it in your app
- If you are upgrading from an earlier version or from Sprockets, this is how you do it
- and finally, if we keep it in: the evolution of asset management (although since Propshaft does use fingerprinting, perhaps this section can be: "Propshaft eliminates the need for practices like the following that you may have encountered"
- This naturally leads to Alternative Libraries and the opening to that section

Reply:
I’ve re-ordered the sections based on your suggestions—thank you! The flow feels much more natural now, and it helped me catch and remove some duplications. I also combined the comparisons of Propshaft and Sprockets into one section instead of spreading them throughout the guide.

Furthermore, I decided to move the "Evolution of Asset Management Techniques" section under "Sprockets to Propshaft," as it felt more fitting there. A lot of the decisions around moving from Sprockets to Propshaft are tied to how asset management techniques have evolved.

I’m almost done addressing all the feedback and plan to push all the updates to a Community PR by later tonight or tomorrow morning. Once it’s up, folks can take a second look there if they'd like.